Bottom: c60b6d50cf4943dfece48059fefc90720891f8bd
Top:    91e680e3edc39fd7db0484c715f591fe8ac91687
Author: Jason Wang <jasowang@redhat.com>
Date:   2017-09-07 20:50:42 +0800

vhost: only fetch avail_idx when necessary

Signed-off-by: Jason Wang <jasowang@redhat.com>


---

diff --git a/drivers/vhost/vhost.c b/drivers/vhost/vhost.c
index 426eb7d..6887b89 100644
--- a/drivers/vhost/vhost.c
+++ b/drivers/vhost/vhost.c
@@ -2492,13 +2492,15 @@ int vhost_prefetch_desc_indices(struct vhost_virtqueue *vq,
 	struct vring_used_elem __user *used;
 	int i;
 
-	if (unlikely(vhost_get_avail(vq, avail_idx, &vq->avail->idx))) {
-		vq_err(vq, "Failed to access avail idx at %p\n",
-		       &vq->avail->idx);
-		return -EFAULT;
+	if (vq->avail_idx == vq->last_avail_idx) {
+		if (unlikely(vhost_get_avail(vq, avail_idx, &vq->avail->idx))) {
+			vq_err(vq, "Failed to access avail idx at %p\n",
+				&vq->avail->idx);
+			return -EFAULT;
+		}
+		vq->avail_idx = vhost16_to_cpu(vq, avail_idx);
 	}
 	last_avail_idx = vq->last_avail_idx & (vq->num - 1);
-	vq->avail_idx = vhost16_to_cpu(vq, avail_idx);
 	total = vq->avail_idx - vq->last_avail_idx;
 	ret = total = min(total, num);
